Refactor Font configuration and import UI, and Font resources [with font size].

bruvzg opened the pull request:
Alternative version of #61406, the same changes but with `font_size` left as a separate theme property.

<img width="478" alt="Screenshot 2022-05-27 at 22 19 18" src="https://user-images.githubusercontent.com/7645683/170778816-809a19a8-5b7a-4905-b4ff-117042917a3a.png"/>

bruvzg committed:
Refactor Font configuration and import UI, and Font resources.

akien-mga commented on doc/classes/Font.xml:
```diff
@@ -247,45 +303,328 @@
+			<argument index="1" name="size" type="Vector2i" />
+			<argument index="2" name="index" type="int" />
+			<description>
+				Renders specified glyph the the font cache texture.
```
```suggestion
				Renders specified glyph to the font cache texture.
```

akien-mga commented on doc/classes/Font.xml:
```diff
@@ -247,45 +303,328 @@
+			<argument index="2" name="texture_index" type="int" />
+			<description>
+				Removes specified texture from font cache entry.
+				[b]Note:[/b] This function will not remove glyphs associated with the texture, remove them manually, using [method remove_glyph].
```
```suggestion
				[b]Note:[/b] This function will not remove glyphs associated with the texture. Remove them manually using [method remove_glyph].
```

akien-mga commented on doc/classes/FontConfig.xml:
```diff
@@ -0,0 +1,279 @@
+		[gdscript]
+		var fc = FontConfig.new()
+		fc.set_font(load("res://BarlowCondensed-Bold.ttf"))
+		fc.set_size(64)
```
Is this correct in this version with separate font size?

bruvzg replied:
No, fixed.

akien-mga commented on doc/classes/FontConfig.xml:
```diff
@@ -0,0 +1,279 @@
+		FontConfig class font configuration used to draw text.
+	</brief_description>
+	<description>
+		[codeblocks]
```
Could use more description before the example code to better explain what FontConfig is used for exactly, and how it differs from and interacts with Font.

bruvzg replied:
Added a bit more usage info to the [Font] and [FontConfig].

bruvzg commented on doc/classes/@GlobalScope.xml:
```diff
@@ -2596,6 +2596,8 @@
 		</constant>
 		<constant name="PROPERTY_USAGE_ARRAY" value="1073741824" enum="PropertyUsageFlags">
 		</constant>
+		<constant name="PROPERTY_USAGE_SHORTCUT" value="-2147483648" enum="PropertyUsageFlags">
```
It was using `uint32_t` to store the values, but I guess enum itself is interpreted as signed (since Variant is casting it to `int64_t`), changed it to `int64_t`. But doc page did not change, there seems to be something else in the doc generator or Variant that still casting it to 32-bit.

fire-forge commented on editor/plugins/font_config_plugin.cpp:
```diff
@@ -0,0 +1,1001 @@
+			prop->update_property();
+		}
+
+		PanelContainer *pc = memnew(PanelContainer);
+		property_vbox->add_child(pc);
+		pc->add_theme_style_override(SNAME("panel"), get_theme_stylebox(SNAME("DictionaryAddItem"), SNAME("EditorStyles")));
+
+		VBoxContainer *add_vbox = memnew(VBoxContainer);
+		pc->add_child(add_vbox);
```
```suggestion
```

These suggestions remove the extra background panel from the OpenType Features editor button. 

Changes it from this:
![image](https://user-images.githubusercontent.com/67974470/173170211-57f6015d-befd-45dd-b324-55a9fe2bb3a6.png)

To look more like this:
![image](https://user-images.githubusercontent.com/67974470/173170283-eed575e6-19a1-4e9e-afae-9588f1e71004.png)

fire-forge commented on editor/plugins/font_config_plugin.cpp:
```diff
@@ -0,0 +1,1001 @@
+			button_add = EditorInspector::create_inspector_action_button(TTR("Add Locale"));
+		}
+		button_add->connect("pressed", callable_mp(this, &EditorPropertyFontMetaOverride::_add_menu));
+		add_vbox->add_child(button_add);
```
```suggestion
		property_vbox->add_child(button_add);
```

fire-forge commented on editor/plugins/font_config_plugin.cpp:
```diff
@@ -0,0 +1,1001 @@
+			}
+		}
+
+		PanelContainer *pc = memnew(PanelContainer);
+		property_vbox->add_child(pc);
+		pc->add_theme_style_override(SNAME("panel"), get_theme_stylebox(SNAME("DictionaryAddItem"), SNAME("EditorStyles")));
+
+		VBoxContainer *add_vbox = memnew(VBoxContainer);
+		pc->add_child(add_vbox);
```
```suggestion
```

fire-forge commented on editor/plugins/font_config_plugin.cpp:
```diff
@@ -0,0 +1,1001 @@
+		button_add = EditorInspector::create_inspector_action_button(TTR("Add Feature"));
+		button_add->set_icon(get_theme_icon(SNAME("Add"), SNAME("EditorIcons")));
+		button_add->connect("pressed", callable_mp(this, &EditorPropertyOTFeatures::_add_menu));
+		add_vbox->add_child(button_add);
```
```suggestion
		property_vbox->add_child(button_add);
```

fire-forge commented on scene/resources/font.cpp:
```diff
@@ -1526,798 +1546,947 @@ void FontData::remove_cache(int p_cache_index) {
+	ADD_PROPERTY(PropertyInfo(Variant::DICTIONARY, "opentype_features"), "set_opentype_features", "get_opentype_features");
+
+	ADD_GROUP("Extra Spacing", "spacing");
+	ADD_PROPERTYI(PropertyInfo(Variant::INT, "spacing_glyph"), "set_spacing", "get_spacing", TextServer::SPACING_GLYPH);
+	ADD_PROPERTYI(PropertyInfo(Variant::INT, "spacing_space"), "set_spacing", "get_spacing", TextServer::SPACING_SPACE);
+	ADD_PROPERTYI(PropertyInfo(Variant::INT, "spacing_top"), "set_spacing", "get_spacing", TextServer::SPACING_TOP);
+	ADD_PROPERTYI(PropertyInfo(Variant::INT, "spacing_bottom"), "set_spacing", "get_spacing", TextServer::SPACING_BOTTOM);
```
```suggestion
	ADD_PROPERTYI(PropertyInfo(Variant::INT, "spacing_glyph", PROPERTY_HINT_NONE, "suffix:px"), "set_spacing", "get_spacing", TextServer::SPACING_GLYPH);
	ADD_PROPERTYI(PropertyInfo(Variant::INT, "spacing_space", PROPERTY_HINT_NONE, "suffix:px"), "set_spacing", "get_spacing", TextServer::SPACING_SPACE);
	ADD_PROPERTYI(PropertyInfo(Variant::INT, "spacing_top", PROPERTY_HINT_NONE, "suffix:px"), "set_spacing", "get_spacing", TextServer::SPACING_TOP);
	ADD_PROPERTYI(PropertyInfo(Variant::INT, "spacing_bottom", PROPERTY_HINT_NONE, "suffix:px"), "set_spacing", "get_spacing", TextServer::SPACING_BOTTOM);
```

Adds suffixes/units to properties (see https://github.com/godotengine/godot/pull/50106, https://github.com/godotengine/godot/pull/61240, and https://github.com/godotengine/godot/pull/50009)

fire-forge commented on scene/resources/font.cpp:
```diff
@@ -1526,798 +1546,947 @@ void FontData::remove_cache(int p_cache_index) {
+	ADD_GROUP("Variation", "variation");
+	ADD_PROPERTY(PropertyInfo(Variant::DICTIONARY, "variation_opentype"), "set_variation_opentype", "get_variation_opentype");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "variation_embolden", PROPERTY_HINT_RANGE, "-2,2,0.01"), "set_variation_embolden", "get_variation_embolden");
+	ADD_PROPERTY(PropertyInfo(Variant::TRANSFORM2D, "variation_transform"), "set_variation_transform", "get_variation_transform");
```
```suggestion
	ADD_PROPERTY(PropertyInfo(Variant::TRANSFORM2D, "variation_transform", PROPERTY_HINT_NONE, "suffix:px"), "set_variation_transform", "get_variation_transform");
```

Suffixes are only added to the `xo` and `yo` editors in Transform2D (which is measured in pixels here).

fire-forge reviewed (commented):
Aside from the minor design changes I commented, I found one bug: text centering doesn't work when changing `FontConfig.spacing_glyph`. Here is a Label with horizontally centered text:

`spacing_glyph` = 0 (correctly centered):
![image](https://user-images.githubusercontent.com/67974470/173170970-53e2178c-3010-41a0-aa29-07bc52ba1af5.png)


`spacing_glyph` = 10 (incorrectly centered):
![image](https://user-images.githubusercontent.com/67974470/173170979-d703fedb-241a-4830-af5f-b5ad2076b85b.png)

The first character moves to the left and the second character stays in the same place. I would expect that the text stays centered.

bruvzg commented:
> Aside from the minor design changes I commented, I found one bug: text centering doesn't work when changing FontConfig.spacing_glyph. Here is a Label with horizontally centered text

Fixed, it was adding extra spacing to the last glyph.
<img width="142" alt="Screenshot 2022-06-11 at 18 02 07" src="https://user-images.githubusercontent.com/7645683/173193415-aef0184b-7fb1-4a86-9598-480c5db2b06c.png">

> These suggestions remove the extra background panel from the OpenType Features editor button.

It looks better indeed, but I have added an extra spacer on the bottom.

| Without spacer | With spacer |
| --- | --- |
| <img width="403" alt="Screenshot 2022-06-11 at 18 05 07" src="https://user-images.githubusercontent.com/7645683/173193417-245571a7-0a9a-4398-a6ac-9227a6698f6c.png"> | <img width="403" alt="Screenshot 2022-06-11 at 18 06 16" src="https://user-images.githubusercontent.com/7645683/173193419-f0861de1-1857-4221-82f4-a1d4aaad35da.png"> |

Probably the same should be done to `Array`/`Dictionary` editors, which use the same code.

fire-forge commented:
> It looks better indeed, but I have added an extra spacer on the bottom.

> Probably the same should be done to `Array`/`Dictionary` editors, which use the same code.

I actually have a PR open (https://github.com/godotengine/godot/pull/59797) that does just that. It is in need of a rebase because it's been sitting for a few months. Depending on which one is merged first, I can apply those changes to the font config editor in that PR or you can apply those changes here.

bruvzg commented:
- Moved "property shortcuts" to the separate PR (independent of this one) - #61991
- Updated to allow selecting face index from TTC / OTC font collections (see #61772).
- Replaced spaces with `MarginContainer4px` container (see #59797).

reduz commented:
I have to say I prefer the version in #61473. I understand @YuriSizov's concern of having to duplicate the resource to change the font size, but this is easily addressed by making the resource unique on duplicate, by adding the hint: PROPERTY_USAGE_DO_NOT_SHARE_ON_DUPLICATE to it. For users it will always be a unique thing.

The font size standalone kind of sucks because its confusing whether you use a theme or not. If its in the resource, then its clear that resource overrides theme and its a simple rule. I also would not make the property checkable, its easier to just see if the resource exists or not and if it exists it overrides theme.

reduz commented:
Regarding implementation wise, there are some things I don't quite get, which are a bit broad to put as code review so here are my doubts:
* I don't understand why Font is being replaced by FontConfig in CanvasItem and most controls. Did we not decide to go against this direction in 4.0? Font on one size and other settings on the other, separate.
* My understanding for FontConfig was more that we need to just have something like FontConfig or FontSettings only for very few specific controls like Label, Label3D or Button because those are controls that users may want to use outside the theme. Having it in places like Tree as an example does not make a lot of sense to me.

bruvzg commented:
> I have to say I prefer the version in https://github.com/godotengine/godot/pull/61406.

I have opened a new PR with font size moved to `FontConfig`, will keep it as two commits (one is the same as this, and second to move font size), since it's easier to rebase both versions this way - #62065

bruvzg commented:
> I don't understand why Font is being replaced by FontConfig in CanvasItem and most controls. Did we not decide to go against this direction in 4.0? Font on one side, and size and other settings on the other, separate.
>
> Having it in places like Tree as an example does not make a lot of sense to me.

Everything in the `FontConfig` is useful for any control or function that's drawing text, there's nothing that's specific to `Label` or other controls. My idea was to move all font related settings from the controls, `Font` and `FontData` to the single resource to be accessible everywhere. So `Font` is what imported for a font file (which might have multiple font faces, for example), and `FontConfig` is a specific variation of the font that's used to actually draw text.

For some controls, it makes sense for font settings more accessible, there's another PR for this - https://github.com/godotengine/godot/pull/61991, but I do not see any reason to use different ways to set the same things.

reduz commented:
@bruvzg I understand this, but the way I see this, its still very annoying because you are forcing the allocation of a FontConfig every time you want to draw something. I would suggest the following reorganization of the font resources:

## **Font**
* Contains what was in FontData
* Contain all the functions for drawing from Font
## **FontVariation**
* Contains variation options, you can pass this to the font drawing code in *Font* optionally, but its not mandatory. You use only if you _really_ need it (which in far most cases you don't). You can expose this as a property in controls that make sense and the controls can just pass it to the draw functions in Font.
## **FontSettings**
* Contains a Font, a size, outline settings, variation, etc. this is *only* for use in classes like Label, Button, Label3D and no more than that. It is of very limited scope meant for classes where you want it to be easier to ignore the theme entirely and use those settings to draw. 
* In those few controls, you need to add some if() condition to check whether this is present and use this to draw rather than the theme. **Entirely hardcoded because that's the point**. 

With this, we have just a single Font class that does everything. A FontVariation class that you can use in the rare use cases where you need this, without forcing it to the user, and a FontSettings class that is only used for some few controls where you want to override and ignore theme.

YuriSizov commented:
> I understand @YuriSizov's concern of having to duplicate the resource to change the font size, but this is easily addressed by making the resource unique on duplicate, by adding the hint: PROPERTY_USAGE_DO_NOT_SHARE_ON_DUPLICATE to it. For users it will always be a unique thing that they can expand and edit if they want, or ignore and use the theme otherwise.

How do we communicate to the user that out of all resources this one cannot be shared for some reason? My main concern is that changing a font size should be just that - changing the size. It's the most frustrating part of working with fonts in 3.x to our users that they have to work with resources to do that. It's not intuitive, not how people treat fonts at all.

Remember, this is a replacement to a toolbar that already meets the expectations of an average user who has ever worked with fonts in any other application. In that toolbar changing font properties, however foreign to Godot's vision it was, was intuitive and familiar. This suggestion complicates things, even if it makes it neater for our internal design and logic. IMO user's expectations should come first and our internal concerns should come second.

So, to reiterate, changing a size of a font is a highly used operation that should be accessible. It should be as simple as changing a numeric value. If you want to have it stored in a resource internally, fine, but it should not require copying of resources on the user side. And as a sidebar, if we do have a flag to implicitly copy a resource instead of sharing, we need to communicate it very well to the user, because this is also not the expected behavior anywhere in Godot.

bruvzg commented:
> You use only if you really need it (which in far most cases you don't). You can expose this as a property in controls that make sense and the controls can just pass it to the draw functions in Font.

I do not see any control that won't need, it's pretty much an essential feature. Absence of it will make any variable font or font collection unusable.

Otherwise, `Font` / `FontVariation` makes sense. But there's an issue with `FontVaraiation`, the set of the available variation properties is depended on the specific font, so it won't be usable as an independent resource. Obviously, it can get required property info when it's part of control and inspector plugin can access control's font, but it seems a bit hacky for me.

I'm not sure what's so special with `Label` and `Button` to deserve its own `FontSettings`, outline doesn't seem like something that need to be changed often, and without it's basically the same as `FontVariation`.

bruvzg commented:
I guess another option would be to do exactly the opposite, keep `Font` and font size theme properties, and add a new `font_variation` Theme property. Which do not need to be a new class and can be just a `Dictionary` (all logic is in the inspector plugin anyway).

Text draw functions stay in the `Font`, and take size and variations as optional arguments.

reduz commented:
@bruvzg The problem to me is that, in practice, a control can use multiple fonts and sizes for different things and mix and match them, but changing the variation is very rare or very context specific so I am not entirely sure that having it as a theme property will be of much use. This is why I suggested only exposing it as a regular property in controls that may make sense to use, or even in Control if you want to override for the whole control.

YuriSizov commented:
Ok, the sharing part we can discuss in a separate proposal, but I still advocate for a better font editor tool that would work intuitively, even if internally it's turtles and duplicated resources all the way down.

bruvzg commented:
> but changing the variation is very rare

Changing variations (and the same for font collections) is just a much more convenient way to have multiple fonts. So if I for example want to use normal and bold fonts, I can have a single variable font file instead of two files (which also will take less space). And I won't need to replace files (or event compile new fonts, if the preferred one is not available by default) if I want to change the thickness of the bold font, for example. Also, if we will ever have support for using system fonts (which I'm planning to add), there are not many options, many of the system fonts are shipped as collections.

Zireael07 commented:
> @YuriSizov IMO this touches on a different problem that I was about to open a proposal for discussion about to get feedback, but the bottom line is that most users are completely unaware that, when duplicating, internal subresources are shared. 

For that reason, we really need an indicator of unique vs shared status, as I asked 2 years ago in https://github.com/godotengine/godot-proposals/issues/904

reduz commented:
@Zireael07 Put this up instead, which is hopefully a simpler solution - https://github.com/godotengine/godot-proposals/issues/4672

reduz commented:
@bruvzg The problem is that, in most cases, users just download font packs that come with the different intensities, so its rare to change the boldness via parameter in general. 

How about this then? **FontVariation** can inherit Font too (or FontBase), and it contains both a Font property and the variation properties. So you can still use it as a variation of an existing font. IMO this should be the simplest way to solve the problem.

bruvzg commented:
> The problem is that, in most cases, users just download font packs that come with the different intensities, so its rare to change the boldness via parameter in general.

Well, I understand it's a less common option, but I do not like an idea of arbitrary restricting a feature to the `Label` and `Button`.

> How about this then? FontVariation can inherit Font too (or FontBase), and it contains both a Font property and the variation properties. So you can still use it as a variation of an existing font. IMO this should be the simplest way to solve the problem.

Hm, I like this option, it should work.

So, something like this:
```
FontBase (abstract class).
↳ Font (imported from a font file, no runtime options).
↳ FontVariation (created by user, basicaly the same as `FontConfig` in this PR).
```
This is probably the most logical way to do it indeed.

reduz commented:
@bruvzg Yeah I think that requires the least code change and hacks and it should just simply work transparently with all the existing code.

reduz commented:
@bruvzg 
> * Simplify current `FontData` + `Font` + bunch of control properties ⇾ to the new `FontBase` + `Font` + `FontVariation`: I'll update this PR (or open a new one) tomorrow.

Sounds good!

> * Resource sharing / duplication: No idea what should be done, but it's not something limited to fonts, so probably should be solved separately.

Definitely out of the scope of this.

> * More convenient access to the font / outline / color for the common controls, like `Label`: so far I do not like anything implemented or proposed:
>   

I think the way it is now in theme is ok because it does what it has to do. The only problem is the cases where you want to have a custom outline. IMO of this we can also add the outline setting to FontVariation, since its kinda the same thing as with Font/FontData.

>   * Current theme properties are hidden in the bottom of inspector and inconvenient to access.
>   * Shortcuts for the theme properties ([Add support for property shortcuts, to display the same property multiple times. #61991](https://github.com/godotengine/godot/pull/61991)) is a bit faster to access, but might cause more confusion, and feels like a dirty hack.
>   * Current toolbar seems to be most user-friendly, but not working nice with small screen sizes, and inconsistent with the rest of editor UI.
>   * `FontSetting` property/class just for a `Label` seems like overkill, and probably won't be any more convenient and less confusing.

**All** the above is solved by FontSetting. I know it may seem counterintuitive, but to me that *is* the right solution. This is what we discussed originally, the actual use cases that you want to solve are:

* Users who *want to  use themes* will go to the theme editor and edit their theme, and touch some theme overrides. This works fine as-is.
* Users who *don't want to use themes* because they use TextureButton, TextureProgress, TextureRect, but still stumble on the theme stuff when using Label or Label3D. For them going to the theme overrides of Label or Button is annoying. FontSettings merely solves this. Its a minor problem with a simple solution.

reduz commented:
@bruvzg I agree its no the cleanest, but its very simple and solves the problem at hand. We can eventually do something better if we figure out what at some point.

bruvzg commented:
https://github.com/godotengine/godot/pull/62108 - a new PR with `FontBase`/`Font`/`FontVariation`. I'll open a separate PR for the `Label`s `FontConfiguration` a bit later (it should be completely independent for other Font changes).

bruvzg commented:
#62139 - a second part, with the settings resource for `Label` and `Button` (done on top of #62108).
